ffi/src/repr_c.rs

```diff
@@ -51,54 +56,138 @@ pub trait CTypeConvert<'itm, S, C>: Sized {
     unsafe fn try_from_repr_c(source: C, store: &'itm mut Self::FfiStore) -> Result<Self>;
 }
 
-/// Provides the capability to use the type as an out-pointer.
-/// This trait mainly exists to be implemented on IR types.
+/// Replaces output type of the wrapper function generated by `ffi_import`. Most notably, the
+/// necessity to replace the output type for a wrapper function arises when returning a reference
+/// that use store during conversion(i.e. that are cloned) or when returning an opaque type reference.
+///
+/// If a type also implements [`Ir`], i.e. has a defined internal representation, a blanket
+/// implementation of [`FfiOutput`] will be provided.
+pub trait COutput<S>: COutPtr<S> {
+    /// Type used instead of the output type in the wrapper function generated by `ffi_import`
+    type ReturnType;
+}
+
+/// Facilitates the use of [`Self`] as out-pointer.
+///
+/// If a type also implements [`Ir`], i.e. has a defined internal representation, a blanket
+/// implementation of [`COutPtr`] will be provided.
 pub trait COutPtr<S>: CType<S> {
     /// Out-pointer type
-    type OutPtr: OutPtrOf<Self::ReprC>;
+    type OutPtr: ReprC;
+}
+
+/// Facilitates writing [`Self`] into [`Self::OutPtr`].
+///
+/// If a type also implements [`Ir`], i.e. has a defined internal representation, a blanket
+/// implementation of [`COutPtr`] will be provided.
+pub trait COutPtrWrite<S>: COutPtr<S> {
+    /// Write the given rust value into the corresponding out-pointer
+    ///
+    /// # Safety
+    ///
+    /// [`*mut Self::OutPtr`] must be valid
+    unsafe fn write_out(self, out_ptr: *mut <Self as COutPtr<S>>::OutPtr);
+}
+
+/// Facilitates reading from [`Self::OutPtr`] out-pointer.
+///
+/// If a type also implements [`Ir`], i.e. has a defined internal representation, a blanket
+/// implementation of [`COutPtr`] will be provided.
+pub trait COutPtrRead<S>: COutPtr<S> + Sized {
+    /// Read a rust value from the corresponding out-pointer
+    ///
+    /// # Errors
+    ///
+    /// Check [`CConvert::try_from_repr_c`]
+    ///
+    /// # Safety
+    ///
+    /// Check [`CConvert::try_from_repr_c`]
+    unsafe fn try_read_out(out_ptr: Self::OutPtr) -> Result<Self>;
 }
 
-/// Marker trait indicating that a type converted into corresponding [`CType`] doesn't
-/// reference conversion context. This is useful to determine which types can be returned
-/// from FFI function and how they are returned.
+/// Marker trait indicating that [`CTypeConvert::into_repr_c`] and [`CTypeConvert::try_from_repr_c`] don't
+/// return a reference to the store. This is useful to determine which(and how) types can be
+/// returned from an FFI function considering that, after return, local context is destroyed
```

Comment:
```suggestion
/// return a reference to the store. 
///
/// Since local context is destroyed after return, this trait is useful in determining which types can be
/// returned from an FFI function.
```